src/Sparrow/Platform/Posix/Syscall.cs

```diff
@@ -333,16 +340,77 @@ public static bool CheckSyncDirectoryAllowed(string filepath)
             return syncAllowed;
         }
 
-        public static int SyncDirectory(string file)
+        public static int SyncDirectory(string file, bool isRealPath = false)
         {
-            var dir = Path.GetDirectoryName(file);
+            var dir = isRealPath == false ? Path.GetDirectoryName(file) : file;
             var fd = open(dir, 0, 0);
             if (fd == -1)
-                return -1;
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to open directory " + dir + " in order to sync for file " + file);
+            }
             var fsyncRc = FSync(fd);
             if (fsyncRc == -1)
-                return -1;
-            return close(fd);
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to FSync directory " + dir + " with fd=" + fd + " in order to sync for file " + file);
+            }
+
+            if (close(fd) == -1)
+            {
+                var err = Marshal.GetLastWin32Error();
+                ThrowLastError(err, "failed to close directory " + dir + " with fd=" + fd + " in order to sync for file " + file);
+            }
+
+            if (isRealPath)
+                return 0;
+
+            // now sync the real path if the above is a symlink
+            var buffSize = 64;
+            var realpathToDir = ArrayPool<byte>.Shared.Rent(buffSize);
+            try
+            {
+                // determine buffSize passed to readlink should be using lstat (but this is not available right now)
+                // so we try with rented array of 1024 and if readlink needs more we double the buffSize until it 
+                // in the right length (TODO: change it to lstat / FileInfo(path).Target in the future when https://github.com/dotnet/corefx/issues/26310 is fixed)
+                int numOfBytes;
+                while (true)
+                {
+                    fixed (byte* pBuff = realpathToDir)
+                    {
+                        numOfBytes = readlink(dir, pBuff, (UIntPtr)buffSize);
+                        var err = Marshal.GetLastWin32Error();
+                        if (numOfBytes == -1)
+                        {
+                            if (err == (int)Errno.EINVAL) // EINVAL when applied on non symlink according to readlink's man page 
+                                return 0;
+                            ThrowLastError(err, "failed to readlink symlink directory " + dir + " with fd=" + fd + " in order to sync for file " + file);
+                        }
+
+                        if (numOfBytes >= buffSize - 1)
```

Comment:
This will never be called. If the buffer is to short, `readlink` will return -1, and you'll never get there.
The return value is the number of bytes _placed_ in the buffer, not the actual size

Reply:
no. it returns "buffSize"

Reply:
I actually tried it with initial buffSize = 8 and path with 30 chars length